src/unix/stream.c

```diff
@@ -666,26 +657,31 @@ static void uv__drain(uv_stream_t* stream) {
   uv_shutdown_t* req;
   int err;
 
+  if (!(stream->flags & UV_HANDLE_SHUTTING))
```

Comment:
Maybe get rid of `UV_HANDLE_SHUTTING`. Seems it's implied when `stream->shutdown_req != NULL`.

`UV_HANDLE_SHUT` I don't really see the use of either. It's to prevent multiple calls to shutdown(2) but I don't see anything wrong with allowing that. The second one will just be a no-op.

Reply:
Those both sound good, but conceptually also separate from this PR. I think it may be best if we merge these 3, then do the change to remove the state bit. I think the conceptual reason to prevent multiple calls is that we can only store one event, so it is slightly more consistent to always prevent queuing a second call, regardless of whether the callback has run for the first call.